### PR TITLE
78 7 day cooldown period on all withdrawals

### PR DESCRIPTION
**Pull request by SamBorisov**
1 - Change the withdrawal period from one Epoch to 7 days
2 - Add a function for governance to change the period of withdrawing
3 - Adjust old tests to work
4 - Add new tests
5 -  Add check to not be able to withdraw if withdrawable is 0
6 - Add and adjust comments for functions

### CODE REVIEW

File: test/RewardPool/RewardPool.test.ts [hunk 1]
```diff
@@ -360,29 +360,62 @@ export function RunVestedDelegateClaimTests(): void {
...
 
+      await vestManager.withdraw(vestManagerOwner.address);
       expect(
         await rewardPool.getRawDelegatorReward(delegatedValidator.address, vestManager.address),
         "getRawDelegatorReward"
       ).to.be.gt(0);
+      expect(await validatorSet.withdrawable(vestManager.address), "withdrawable").to.be.eq(0);
+    });
+
+    it("should return when active position & claim", async function () {
```
Comment (R-Santev) [inline test/RewardPool/RewardPool.test.ts:399]: What is the purpose of this test?

**Review by R-Santev — CHANGES_REQUESTED** [review]

File: contracts/ValidatorSet/libs/WithdrawalQueue.sol [hunk 2]
```diff
@@ -12,35 +12,26 @@ import "./IWithdrawalQueue.sol";
...
             return;
         }
 
-        uint256 latestEpoch = self.withdrawals[tail - 1].epoch;
-        assert(epoch >= latestEpoch);
-        if (latestEpoch < epoch) {
-            // new withdrawal for next epoch
-            self.withdrawals[tail] = WithdrawalData(amount, epoch);
-            self.tail++;
-        } else {
```
[the remark this answers is not in the document]
Comment (SamBorisov) [reply contracts/ValidatorSet/libs/WithdrawalQueue.sol:40]: fixed

File: contracts/ValidatorSet/modules/Withdrawal/IWithdrawal.sol [hunk 3]
```diff
@@ -12,7 +12,7 @@ interface IWithdrawal {
     function withdraw(address to) external;
 
     /**
-     * @notice Calculates how much can be withdrawn for account in this epoch.
+     * @notice Calculates how much can be withdrawn for account in this time.
```
[the remark this answers is not in the document]
Comment (SamBorisov) [reply contracts/ValidatorSet/modules/Withdrawal/IWithdrawal.sol]: fixed

File: contracts/ValidatorSet/modules/Withdrawal/Withdrawal.sol [hunk 4]
```diff
@@ -21,7 +21,8 @@ abstract contract Withdrawal is IWithdrawal, ReentrancyGuardUpgradeable, Validat
     function withdraw(address to) external nonReentrant {
         assert(to != address(0));
         WithdrawalQueue storage queue = _withdrawals[msg.sender];
-        (uint256 amount, uint256 newHead) = queue.withdrawable(currentEpochId);
+        (uint256 amount, uint256 newHead) = queue.withdrawable();
+        require(amount > 0, "NO_WITHDRAWAL_AVAILABLE");
```
[the remark this answers is not in the document]
Comment (SamBorisov) [reply contracts/ValidatorSet/modules/Withdrawal/Withdrawal.sol]: added custom error

File: contracts/ValidatorSet/modules/Withdrawal/Withdrawal.sol [hunk 5]
```diff
@@ -35,18 +36,27 @@ abstract contract Withdrawal is IWithdrawal, ReentrancyGuardUpgradeable, Validat
...
     function pendingWithdrawals(address account) external view returns (uint256) {
-        return _withdrawals[account].pending(currentEpochId);
+        return _withdrawals[account].pending();
+    }
+
+    /**
+     * @inheritdoc IWithdrawal
+     */
+    function changeWithdrawalWaitPeriod(uint256 newWaitPeriod) external onlyOwner {
+        require(newWaitPeriod > 0, "INVALID_WAIT_PERIOD");
```
[the remark this answers is not in the document]
Comment (SamBorisov) [reply contracts/ValidatorSet/modules/Withdrawal/Withdrawal.sol]: added

File: contracts/ValidatorSet/modules/Withdrawal/IWithdrawal.sol [hunk 6]
```diff
@@ -24,4 +24,11 @@ interface IWithdrawal {
      * @return Amount not yet withdrawable (in wei)
      */
     function pendingWithdrawals(address account) external view returns (uint256);
+
+    /**
+     * @notice Changes the withdrawal wait period.
+     * @dev This function should be called only by the Governed contract.
+     * @param newWaitPeriod The new withdrawal wait period
```
[the remark this answers is not in the document]
Comment (SamBorisov) [reply contracts/ValidatorSet/modules/Withdrawal/IWithdrawal.sol]: Added more text

File: test/ValidatorSet/Delegation.test.ts [hunk 7]
```diff
@@ -525,13 +526,8 @@ export function RunDelegationTests(): void {
...
         const balanceBefore = await vestManagerOwner.getBalance();
 
-        // commit Epoch so reward is available for withdrawal
-        await commitEpoch(
-          systemValidatorSet,
-          rewardPool,
-          [this.signers.validators[0], this.signers.validators[1], delegatedValidator],
-          this.epochSize
-        );
+        // commit increase time so reward is available for withdrawal
```
[the remark this answers is not in the document]
Comment (SamBorisov) [reply test/ValidatorSet/Delegation.test.ts]: changed

File: test/RewardPool/RewardPool.test.ts [hunk 8]
```diff
@@ -360,29 +360,62 @@ export function RunVestedDelegateClaimTests(): void {
       ).to.be.revertedWith("Ownable: caller is not the owner");
     });
 
-    it("should return when active position", async function () {
-      const { systemValidatorSet, validatorSet, rewardPool, vestManager, vestManagerOwner, delegatedValidator } =
-        await loadFixture(this.fixtures.weeklyVestedDelegationFixture);
+    it("should return when active position & withdraw", async function () {
```
[the remark this answers is not in the document]
Comment (SamBorisov) [reply test/RewardPool/RewardPool.test.ts:363]: removed withdraw & added checks if the balance stays the same before and after claiming

File: test/RewardPool/RewardPool.test.ts (hunk 1 above)
Comment (SamBorisov) [reply test/RewardPool/RewardPool.test.ts:399]: apply needed checks in previous test and removed this one

File: contracts/ValidatorSet/modules/Withdrawal/Withdrawal.sol [hunk 9]
```diff
@@ -8,6 +8,9 @@ import "./../AccessControl/AccessControl.sol";
 import "./../../ValidatorSetBase.sol";
 import "./../../libs/WithdrawalQueue.sol";
 
+error NoWithdrawalAvailable();
```
Comment (R-Santev) [inline contracts/ValidatorSet/modules/Withdrawal/Withdrawal.sol:11]: Custom errors that are module-specific to be declared in the module's interface or in the module itself if the interface is not available. (internal styling convention)

File: test/RewardPool/RewardPool.test.ts [hunk 10]
```diff
@@ -360,51 +360,17 @@ export function RunVestedDelegateClaimTests(): void {
...
-      await vestManager.withdraw(vestManagerOwner.address);
-      expect(
-        await rewardPool.getRawDelegatorReward(delegatedValidator.address, vestManager.address),
-        "getRawDelegatorReward"
-      ).to.be.gt(0);
-      expect(await validatorSet.withdrawable(vestManager.address), "withdrawable").to.be.eq(0);
-    });
-
-    it("should return when active position & claim", async function () {
+    it("should return when active position", async function () {
```
Comment (R-Santev) [inline test/RewardPool/RewardPool.test.ts:363]:
```suggestion
    it("should not claim when active position", async function () {
```

File: test/RewardPool/RewardPool.test.ts [hunk 11]
```diff
@@ -416,9 +382,13 @@ export function RunVestedDelegateClaimTests(): void {
         await rewardPool.getRawDelegatorReward(delegatedValidator.address, vestManager.address),
         "getRawDelegatorReward"
       ).to.be.gt(0);
+
       // claim
       await vestManager.claimVestedPositionReward(delegatedValidator.address, 0, 0);
       expect(await validatorSet.withdrawable(vestManager.address), "withdrawable").to.be.eq(0);
```
Comment (R-Santev) [inline test/RewardPool/RewardPool.test.ts:388]: This check is not needed

**Review by R-Santev — CHANGES_REQUESTED** [review]

File: contracts/ValidatorSet/modules/Withdrawal/Withdrawal.sol (hunk 9 above)
Comment (SamBorisov) [reply contracts/ValidatorSet/modules/Withdrawal/Withdrawal.sol]: Moved to the interface

File: test/RewardPool/RewardPool.test.ts (hunk 11 above)
Comment (SamBorisov) [reply test/RewardPool/RewardPool.test.ts]: removed

**Pull request merged**